nomos-mix/network/src/behaviour.rs

```diff
@@ -127,18 +163,118 @@ where
+                    }
+                    removed
+                }
+                None => false,
+            },
+            None => self.negotiated_peers.remove(peer_id).is_some(),
+        }
+    }
+
+    fn is_negotiated_peer(&self, peer_id: &PeerId) -> bool {
+        self.negotiated_peers.contains_key(peer_id)
+    }
+
+    /// Handle newly detected malicious and unhealthy peers and schedule new connection establishments.
+    /// This must be called at the end of each time window.
+    fn run_conn_maintenance(&mut self) {
+        if let Some(conn_maintenance) = self.conn_maintenance.as_mut() {
+            let (malicious_peers, unhealthy_peers) = conn_maintenance.reset();
+            let num_closed_malicious = self.handle_malicious_peers(malicious_peers);
+            let num_connected_unhealthy = self.handle_unhealthy_peers(unhealthy_peers);
+            let mut num_to_dial = num_closed_malicious + num_connected_unhealthy;
+            if num_to_dial + self.negotiated_peers.len() > self.config.max_peering_degree {
+                tracing::warn!(
+                    "Cannot establish {} new connections due to max_peering_degree:{}. connected_peers:{}",
+                    num_to_dial,
+                    self.config.max_peering_degree,
+                    self.negotiated_peers.len(),
+                );
+                num_to_dial = self.config.max_peering_degree - self.negotiated_peers.len();
             }
+            self.schedule_dials(num_to_dial);
+        }
+    }
+
+    /// Add newly detected malicious peers to the blacklist,
+    /// and schedule connection closures if the connections exist.
+    /// This returns the number of connection closures that were actually scheduled.
+    fn handle_malicious_peers(&mut self, peers: HashSet<PeerId>) -> usize {
+        peers
+            .into_iter()
+            .map(|peer_id| {
+                self.blacklist_peers.insert(peer_id);
+                // Close the connection only if the peer was already connected.
+                if self.remove_negotiated_peer(&peer_id, None) {
+                    self.events.push_back(ToSwarm::CloseConnection {
+                        peer_id,
+                        connection: libp2p::swarm::CloseConnection::All,
+                    });
+                    1
+                } else {
+                    0
+                }
+            })
+            .sum()
+    }
+
+    /// Add newly detected unhealthy peers to the blacklist,
+    /// and return the number of connections that actually exist to the unhealthy peers.
+    fn handle_unhealthy_peers(&mut self, peers: HashSet<PeerId>) -> usize {
+        peers
+            .into_iter()
+            .map(|peer_id| {
+                self.blacklist_peers.insert(peer_id);
+                if self.is_negotiated_peer(&peer_id) {
+                    1
+                } else {
+                    0
+                }
+            })
+            .sum()
+    }
```

Comment:
I've thought about it but actually it's not easy to abstract this clearly into the `nomos-mix` crate, because the user can establish/close connections in the libp2p swarm level, which means that it is hard for the `nomos-mix` crate to calculate exactly how many new connections should be established.
But anyway, I'll try to refactor this.;

Reply:
I've thought so at the beginning, but it was not easy to abstract it into the `nomos-mix/core`. Connection-related code (not even for connection maintenance) is scattered across mix-service and mix-behaviour. And, they are conceptually coupled with each other closely. 
However, I just found that the specification has been restructured, and now all connection-related logic is defined under the connection maintenance part. Then, I can abstract all connection-related code into `noms-mix/core`, and make mix-service and mix-behaviour thinner.
For that, there is something to be more clarified in the specification. I will discuss with Marcin first and modify this PR.

Reply:
@danielSanchezQ I refactored the entire code base and this PR is now ready for review. Now all connection maintenance logic is in the `nomos-mix` crate. The `nomos-mix-network` crate and `nomos-mix-service` crate are thin.